File: src/main/java/org/cubexmc/metro/train/TrainDisplayController.java
package org.cubexmc.metro.train;

import java.util.List;
import java.util.Map;

import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.EventPriority;

import org.cubexmc.metro.Metro;
import org.cubexmc.metro.event.MetroTrainArrivalEvent;
import org.cubexmc.metro.event.MetroTrainDepartureEvent;
import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.manager.StopManager;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.model.Stop;
import org.cubexmc.metro.util.SchedulerUtil;
import org.cubexmc.metro.util.SoundUtil;
import org.cubexmc.metro.util.TextUtil;

import net.md_5.bungee.api.ChatMessageType;
import net.md_5.bungee.api.chat.TextComponent;

public class TrainDisplayController implements Listener {

    private final Metro plugin;

    public TrainDisplayController(Metro plugin) {
        this.plugin = plugin;
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onTrainArrival(MetroTrainArrivalEvent event) {
        Player passenger = event.getPassenger();
        if (passenger == null || !passenger.isOnline())
            return;

        Stop targetStop = event.getCurrentStop();
        Line line = event.getLine();
        Minecart minecart = event.getMinecart();

        if (event.getArrivalType() == MetroTrainArrivalEvent.ArrivalType.ENTERING) {
            // 播放到站音乐
            playArrivalSound(passenger);
            // 播放站台到站音乐（给站台上的玩家听）
            playStationArrivalSound(targetStop, passenger);

            if (event.isTerminus()) {
                if (plugin.getConfigFacade().isTerminalStopTitleEnabled()) {
                    showTerminalStopInfo(passenger, targetStop, line);
                }
            } else {
                showArriveStopInfo(passenger, targetStop, line);
            }
        } else if (event.getArrivalType() == MetroTrainArrivalEvent.ArrivalType.DOCKED) {
            if (!event.isTerminus()) {
                // 非终点站，显示等待发车信息
                showWaitingInfo(passenger, minecart, targetStop, line);
                // 播放等待发车音乐
                startWaitingSound(minecart, passenger);
            }
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onTrainDeparture(MetroTrainDepartureEvent event) {
        Player passenger = event.getPassenger();
        if (passenger == null || !passenger.isOnline())
            return;

        Stop currentStop = event.getCurrentStop();
        Stop nextStop = event.getNextStop();
        Line line = event.getLine();
        Minecart minecart = event.getMinecart();

        // 播放发车音乐
        playDepartureSound(passenger);

        // 显示行程信息
        if (plugin.getConfigFacade().isDepartureTitleEnabled()) {
            showDepartureInfo(passenger, minecart, currentStop, nextStop, line);
        }
    }

    private void playStationArrivalSound(Stop stop, Player passenger) {
        if (!plugin.getConfigFacade().isStationArrivalSoundEnabled())
            return;

        List<String> notes = plugin.getConfigFacade().getStationArrivalNotes();
        if (notes.isEmpty())
            return;

        int initialDelay = plugin.getConfigFacade().getStationArrivalInitialDelay();
        Location stopLocation = stop.getStopPointLocation();
        if (stopLocation == null || stopLocation.getWorld() == null)
            return;

        for (Player player : stopLocation.getWorld().getPlayers()) {
            if (player.equals(passenger))
                continue;
            if (stop.isInStop(player.getLocation())) {
                SoundUtil.playNoteSequence(plugin, player, notes, initialDelay);
            }
        }
    }

    private void playArrivalSound(Player passenger) {
        if (plugin.getConfigFacade().isArrivalSoundEnabled() && !plugin.getConfigFacade().getArrivalNotes().isEmpty()
                && passenger != null) {
            SoundUtil.playNoteSequence(plugin, passenger, plugin.getConfigFacade().getArrivalNotes(),
                    plugin.getConfigFacade().getArrivalInitialDelay());
        }
    }

    private void playDepartureSound(Player passenger) {
        if (plugin.getConfigFacade().isDepartureSoundEnabled()
                && !plugin.getConfigFacade().getDepartureNotes().isEmpty() && passenger != null) {
            SoundUtil.playNoteSequence(plugin, passenger, plugin.getConfigFacade().getDepartureNotes(),
                    plugin.getConfigFacade().getDepartureInitialDelay());
        }
    }

    private void showArriveStopInfo(Player passenger, Stop stop, Line line) {
        if (passenger == null || !passenger.isOnline())
            return;

        String nextStopId = line.getNextStopId(stop.getId());
        Stop nextStop = nextStopId != null ? plugin.getStopManager().getStop(nextStopId) : null;

        List<String> stopIds = line.getOrderedStopIds();
        Stop terminusStop = !stopIds.isEmpty() ? plugin.getStopManager().getStop(stopIds.get(stopIds.size() - 1))
                : null;

        showStopInfo(passenger, line, "arrive_stop", stop, null, nextStop, terminusStop,
                plugin.getConfigFacade().getArriveStopFadeIn(), plugin.getConfigFacade().getArriveStopStay(),
                plugin.getConfigFacade().getArriveStopFadeOut());
    }

    private void showTerminalStopInfo(Player passenger, Stop stop, Line line) {
        if (passenger == null || !passenger.isOnline())
            return;

        int fadeIn = plugin.getConfigFacade().getTerminalStopFadeIn();
        int stay = plugin.getConfigFacade().getTerminalStopStay();
        int fadeOut = plugin.getConfigFacade().getTerminalStopFadeOut();

        showStopInfo(passenger, line, "terminal_stop", stop, null, null, stop, fadeIn, stay, fadeOut);
    }

    private void showDepartureInfo(Player passenger, Minecart minecart, Stop currentStop, Stop nextStop, Line line) {
        if (passenger == null || !passenger.isOnline())
            return;

        List<String> stopIds = line.getOrderedStopIds();
        Stop terminusStop = !stopIds.isEmpty() ? plugin.getStopManager().getStop(stopIds.get(stopIds.size() - 1))
                : null;

        int fadeIn = plugin.getConfigFacade().getDepartureFadeIn();
        int stay = plugin.getConfigFacade().getDepartureStay();
        int fadeOut = plugin.getConfigFacade().getDepartureFadeOut();

        String actionbarTemplate = plugin.getConfigFacade().getDepartureActionbar();
        Map<String, String> customTitle = currentStop.getCustomTitle("departure");
        if (customTitle != null && customTitle.containsKey("actionbar")) {
            actionbarTemplate = customTitle.get("actionbar");
        }

        if (actionbarTemplate.contains("{countdown}")) {
            startCountdownActionbar(passenger, minecart, line, "departure", currentStop, currentStop, nextStop,
                    terminusStop);
        }

        showStopInfo(passenger, line, "departure", currentStop, currentStop, nextStop, terminusStop, fadeIn, stay,
                fadeOut);
    }

    private void showWaitingInfo(Player passenger, Minecart minecart, Stop currentStop, Line line) {
        if (passenger == null || !passenger.isOnline())
            return;
        if (!plugin.getConfigFacade().isWaitingTitleEnabled())
            return;

        StopManager stopManager = plugin.getStopManager();
        String nextStopId = line.getNextStopId(currentStop.getId());
        Stop nextStop = nextStopId != null ? stopManager.getStop(nextStopId) : null;

        List<String> stopIds = line.getOrderedStopIds();
        Stop terminusStop = !stopIds.isEmpty() ? stopManager.getStop(stopIds.get(stopIds.size() - 1)) : null;

        LineManager lineManager = plugin.getLineManager();

        String titleTemplate = plugin.getConfigFacade().getWaitingTitle();
        String subtitleTemplate = plugin.getConfigFacade().getWaitingSubtitle();

        Map<String, String> customTitle = currentStop.getCustomTitle("waiting");
        if (customTitle != null) {
            if (customTitle.containsKey("title"))
                titleTemplate = customTitle.get("title");
            if (customTitle.containsKey("subtitle"))
                subtitleTemplate = customTitle.get("subtitle");
        }

        String title = TextUtil.replacePlaceholders(titleTemplate, line, currentStop, null, nextStop, terminusStop,
                lineManager);
        String subtitle = TextUtil.replacePlaceholders(subtitleTemplate, line, currentStop, null, nextStop,
                terminusStop, lineManager);

        title = ChatColor.translateAlternateColorCodes('&', title);
        subtitle = ChatColor.translateAlternateColorCodes('&', subtitle);

        passenger.sendTitle(title, subtitle, 0, 1000000, 0);

        String actionbarTemplate = plugin.getConfigFacade().getWaitingActionbar();
        if (customTitle != null && customTitle.containsKey("actionbar")) {
            actionbarTemplate = customTitle.get("actionbar");
        }

        if (actionbarTemplate.contains("{countdown}")) {
            startCountdownActionbar(passenger, minecart, line, "waiting", currentStop, null, nextStop, terminusStop);
        } else {
            String actionbarText = TextUtil.replacePlaceholders(actionbarTemplate, line, currentStop, null, nextStop,
                    terminusStop, lineManager);
            actionbarText = ChatColor.translateAlternateColorCodes('&', actionbarText);
            passenger.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent.fromLegacyText(actionbarText));
        }
    }

    private void startCountdownActionbar(Player passenger, Minecart minecart, Line line, String infoType,
            Stop mainStop, Stop prevStop, Stop nextStop, Stop terminusStop) {
        if (passenger == null || !passenger.isOnline())
            return;

        String actionbarTemplate = "";
        switch (infoType) {
            case "waiting":
                actionbarTemplate = plugin.getConfigFacade().getWaitingActionbar();
                break;
            case "departure":
                actionbarTemplate = plugin.getConfigFacade().getDepartureActionbar();
                break;
            case "arrive_stop":
            case "terminal_stop":
                // No default actionbar for these yet, wait for implementation if needed
                break;
        }

        Map<String, String> customTitle = mainStop.getCustomTitle(infoType);
        if (customTitle != null && customTitle.containsKey("actionbar")) {
            actionbarTemplate = customTitle.get("actionbar");
        }

        final String template = actionbarTemplate;
        LineManager lineManager = plugin.getLineManager();
        long durationTicks = "waiting".equals(infoType)
                ? getWaitingDisplayDurationTicks(minecart)
                : plugin.getConfigFacade().getCartDepartureDelay();
        int totalSeconds = (int) Math.ceil(durationTicks / 20.0);

        for (int i = totalSeconds; i >= 0; i--) {
            final int secondsLeft = i;
            long delayTicks = (totalSeconds - i) * 20L;

            scheduleTrainTask(minecart, () -> {
                if (!isPassengerStillRidingTrain(minecart, passenger))
                    return;

                String text = template.replace("{countdown}", String.valueOf(secondsLeft));
                text = TextUtil.replacePlaceholders(text, line, mainStop, prevStop, nextStop, terminusStop,
                        lineManager);
                text = ChatColor.translateAlternateColorCodes('&', text);

                passenger.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent.fromLegacyText(text));
            }, delayTicks, -1);
        }
    }

    private void startWaitingSound(Minecart minecart, Player passenger) {
        if (!plugin.getConfigFacade().isWaitingSoundEnabled() || plugin.getConfigFacade().getWaitingNotes().isEmpty()
                || passenger == null)
            return;

        scheduleTrainTask(minecart, () -> {
            if (isPassengerStillRidingTrain(minecart, passenger)) {
                playWaitingSoundOnce(passenger);
            }
        }, plugin.getConfigFacade().getWaitingInitialDelay(), -1);

        int interval = plugin.getConfigFacade().getWaitingSoundInterval();
        if (interval <= 0)
            return;

        long durationTicks = getWaitingDisplayDurationTicks(minecart);
        long repeatTimes = (durationTicks + interval - 1L) / interval;

        for (long i = 1; i <= repeatTimes; i++) {
            scheduleTrainTask(minecart, () -> {
                if (isPassengerStillRidingTrain(minecart, passenger)) {
                    playWaitingSoundOnce(passenger);
                }
            }, plugin.getConfigFacade().getWaitingInitialDelay() + (interval * i), -1);
        }
    }

    private long getWaitingDisplayDurationTicks(Minecart minecart) {
        TrainInstance serviceTrain = findServiceTrain(minecart);
        if (serviceTrain != null && serviceTrain.isWaiting()) {
            return serviceTrain.getRemainingDwellTicks(SchedulerUtil.getCurrentTick());
        }
        return plugin.getConfigFacade().getCartDepartureDelay();
    }

    private boolean isPassengerStillRidingTrain(Minecart minecart, Player passenger) {
        if (passenger == null || !passenger.isOnline()) {
            return false;
        }
        if (passenger.getVehicle() == minecart) {
            return true;
        }
        TrainInstance serviceTrain = findServiceTrain(minecart);
        return serviceTrain != null && serviceTrain.isPassenger(passenger);
    }

    private TrainInstance findServiceTrain(Minecart minecart) {
        if (minecart == null || plugin.getLineServiceManager() == null) {
            return null;
        }
        return plugin.getLineServiceManager().getTrainByMinecart(minecart.getUniqueId());
    }

    private Object scheduleTrainTask(Minecart minecart, Runnable task, long delay, long period) {
        TrainMovementTask movementTask = TrainMovementTask.getTaskFor(minecart);
        if (movementTask != null) {
            return movementTask.scheduleSessionTask(task, delay, period);
        }
        return SchedulerUtil.entityRun(plugin, minecart, task, delay, period);
    }

    private void playWaitingSoundOnce(Player passenger) {
        if (passenger != null && passenger.isOnline()) {
            SoundUtil.playNoteSequence(plugin, passenger, plugin.getConfigFacade().getWaitingNotes(), 0);
        }
    }

    private void showStopInfo(Player passenger, Line line, String infoType, Stop mainStop, Stop prevStop,
            Stop nextStop, Stop terminusStop, int fadeIn, int stay, int fadeOut) {
        if (passenger == null || !passenger.isOnline() || mainStop == null)
            return;

        LineManager lineManager = plugin.getLineManager();
        String titleTemplate = "";
        String subtitleTemplate = "";
        String actionbarTemplate = "";

        switch (infoType) {
            case "arrive_stop":
                titleTemplate = plugin.getConfigFacade().getArriveStopTitle();
                subtitleTemplate = plugin.getConfigFacade().getArriveStopSubtitle();
                break;
            case "terminal_stop":
                titleTemplate = plugin.getConfigFacade().getTerminalStopTitle();
                subtitleTemplate = plugin.getConfigFacade().getTerminalStopSubtitle();
                break;
            case "departure":
                titleTemplate = plugin.getConfigFacade().getDepartureTitle();
                subtitleTemplate = plugin.getConfigFacade().getDepartureSubtitle();
                actionbarTemplate = plugin.getConfigFacade().getDepartureActionbar();
                break;
            case "waiting":
                titleTemplate = plugin.getConfigFacade().getWaitingTitle();
                subtitleTemplate = plugin.getConfigFacade().getWaitingSubtitle();
                actionbarTemplate = plugin.getConfigFacade().getWaitingActionbar();
                break;
        }

        Map<String, String> customTitle = mainStop.getCustomTitle(infoType);
        if (customTitle != null) {
            if (customTitle.containsKey("title"))
                titleTemplate = customTitle.get("title");
            if (customTitle.containsKey("subtitle"))
                subtitleTemplate = customTitle.get("subtitle");
            if (customTitle.containsKey("actionbar"))
                actionbarTemplate = customTitle.get("actionbar");
        }

        String title = TextUtil.replacePlaceholders(titleTemplate, line, mainStop, prevStop, nextStop, terminusStop,
                lineManager);
        String subtitle = TextUtil.replacePlaceholders(subtitleTemplate, line, mainStop, prevStop, nextStop,
                terminusStop, lineManager);
        String actionbar = TextUtil.replacePlaceholders(actionbarTemplate, line, mainStop, prevStop, nextStop,
                terminusStop, lineManager);

        title = ChatColor.translateAlternateColorCodes('&', title);
        subtitle = ChatColor.translateAlternateColorCodes('&', subtitle);
        actionbar = ChatColor.translateAlternateColorCodes('&', actionbar);

        passenger.sendTitle(title, subtitle, fadeIn, stay, fadeOut);
        passenger.spigot().sendMessage(ChatMessageType.ACTION_BAR, TextComponent.fromLegacyText(actionbar));
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainDisplayController.kt
package org.cubexmc.metro.train

import kotlin.math.ceil
import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.entity.Minecart
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.cubexmc.metro.Metro
import org.cubexmc.metro.event.MetroTrainArrivalEvent
import org.cubexmc.metro.event.MetroTrainDepartureEvent
import org.cubexmc.metro.manager.LineManager
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.util.SchedulerUtil
import org.cubexmc.metro.util.SoundUtil
import org.cubexmc.metro.util.TextUtil

class TrainDisplayController(private val plugin: Metro) : Listener {
    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onTrainArrival(event: MetroTrainArrivalEvent) {
        val passenger = event.passenger
        if (passenger == null || !passenger.isOnline) {
            return
        }

        val targetStop = event.currentStop
        val line = event.line
        val minecart = event.minecart

        if (event.arrivalType == MetroTrainArrivalEvent.ArrivalType.ENTERING) {
            playArrivalSound(passenger)
            playStationArrivalSound(targetStop, passenger)

            if (event.isTerminus()) {
                if (plugin.configFacade.isTerminalStopTitleEnabled()) {
                    showTerminalStopInfo(passenger, targetStop, line)
                }
            } else {
                showArriveStopInfo(passenger, targetStop, line)
            }
        } else if (event.arrivalType == MetroTrainArrivalEvent.ArrivalType.DOCKED) {
            if (!event.isTerminus()) {
                showWaitingInfo(passenger, minecart, targetStop, line)
                startWaitingSound(minecart, passenger)
            }
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onTrainDeparture(event: MetroTrainDepartureEvent) {
        val passenger = event.passenger
        if (passenger == null || !passenger.isOnline) {
            return
        }

        val currentStop = event.currentStop
        val nextStop = event.nextStop
        val line = event.line
        val minecart = event.minecart

        playDepartureSound(passenger)
        if (plugin.configFacade.isDepartureTitleEnabled()) {
            showDepartureInfo(passenger, minecart, currentStop, nextStop, line)
        }
    }

    private fun playStationArrivalSound(stop: Stop, passenger: Player) {
        if (!plugin.configFacade.isStationArrivalSoundEnabled()) {
            return
        }

        val notes = plugin.configFacade.getStationArrivalNotes()
        if (notes.isEmpty()) {
            return
        }

        val initialDelay = plugin.configFacade.getStationArrivalInitialDelay()
        val stopLocation = stop.stopPointLocation
        val world = stopLocation?.world ?: return

        for (player in world.players) {
            if (player == passenger) {
                continue
            }
            if (stop.isInStop(player.location)) {
                SoundUtil.playNoteSequence(plugin, player, notes, initialDelay)
            }
        }
    }

    private fun playArrivalSound(passenger: Player?) {
        if (
            plugin.configFacade.isArrivalSoundEnabled() &&
            plugin.configFacade.getArrivalNotes().isNotEmpty() &&
            passenger != null
        ) {
            SoundUtil.playNoteSequence(
                plugin,
                passenger,
                plugin.configFacade.getArrivalNotes(),
                plugin.configFacade.getArrivalInitialDelay(),
            )
        }
    }

    private fun playDepartureSound(passenger: Player?) {
        if (
            plugin.configFacade.isDepartureSoundEnabled() &&
            plugin.configFacade.getDepartureNotes().isNotEmpty() &&
            passenger != null
        ) {
            SoundUtil.playNoteSequence(
                plugin,
                passenger,
                plugin.configFacade.getDepartureNotes(),
                plugin.configFacade.getDepartureInitialDelay(),
            )
        }
    }

    private fun showArriveStopInfo(passenger: Player?, stop: Stop, line: Line) {
        if (passenger == null || !passenger.isOnline) {
            return
        }

        val nextStopId = line.getNextStopId(stop.id)
        val nextStop = if (nextStopId != null) plugin.stopManager.getStop(nextStopId) else null
        val stopIds = line.orderedStopIds
        val terminusStop = if (stopIds.isNotEmpty()) {
            plugin.stopManager.getStop(stopIds[stopIds.size - 1])
        } else {
            null
        }

        showStopInfo(
            passenger,
            line,
            "arrive_stop",
            stop,
            null,
            nextStop,
            terminusStop,
            plugin.configFacade.getArriveStopFadeIn(),
            plugin.configFacade.getArriveStopStay(),
            plugin.configFacade.getArriveStopFadeOut(),
        )
    }

    private fun showTerminalStopInfo(passenger: Player?, stop: Stop, line: Line) {
        if (passenger == null || !passenger.isOnline) {
            return
        }

        showStopInfo(
            passenger,
            line,
            "terminal_stop",
            stop,
            null,
            null,
            stop,
            plugin.configFacade.getTerminalStopFadeIn(),
            plugin.configFacade.getTerminalStopStay(),
            plugin.configFacade.getTerminalStopFadeOut(),
        )
    }

    private fun showDepartureInfo(
        passenger: Player?,
        minecart: Minecart?,
        currentStop: Stop,
        nextStop: Stop,
        line: Line,
    ) {
        if (passenger == null || !passenger.isOnline) {
            return
        }

        val stopIds = line.orderedStopIds
        val terminusStop = if (stopIds.isNotEmpty()) {
            plugin.stopManager.getStop(stopIds[stopIds.size - 1])
        } else {
            null
        }

        var actionbarTemplate = plugin.configFacade.getDepartureActionbar()
        val customTitle = currentStop.getCustomTitle("departure")
        if (customTitle != null && customTitle.containsKey("actionbar")) {
            actionbarTemplate = customTitle.getValue("actionbar")
        }

        if (actionbarTemplate.contains("{countdown}")) {
            startCountdownActionbar(
                passenger,
                minecart,
                line,
                "departure",
                currentStop,
                currentStop,
                nextStop,
                terminusStop,
            )
        }

        showStopInfo(
            passenger,
            line,
            "departure",
            currentStop,
            currentStop,
            nextStop,
            terminusStop,
            plugin.configFacade.getDepartureFadeIn(),
            plugin.configFacade.getDepartureStay(),
            plugin.configFacade.getDepartureFadeOut(),
        )
    }

    private fun showWaitingInfo(passenger: Player?, minecart: Minecart?, currentStop: Stop, line: Line) {
        if (passenger == null || !passenger.isOnline) {
            return
        }
        if (!plugin.configFacade.isWaitingTitleEnabled()) {
            return
        }

        val stopManager = plugin.stopManager
        val nextStopId = line.getNextStopId(currentStop.id)
        val nextStop = if (nextStopId != null) stopManager.getStop(nextStopId) else null
        val stopIds = line.orderedStopIds
        val terminusStop = if (stopIds.isNotEmpty()) stopManager.getStop(stopIds[stopIds.size - 1]) else null
        val lineManager = plugin.lineManager

        var titleTemplate = plugin.configFacade.getWaitingTitle()
        var subtitleTemplate = plugin.configFacade.getWaitingSubtitle()

        val customTitle = currentStop.getCustomTitle("waiting")
        if (customTitle != null) {
            if (customTitle.containsKey("title")) {
                titleTemplate = customTitle.getValue("title")
            }
            if (customTitle.containsKey("subtitle")) {
                subtitleTemplate = customTitle.getValue("subtitle")
            }
        }

        var title = TextUtil.replacePlaceholders(
            titleTemplate,
            line,
            currentStop,
            null,
            nextStop,
            terminusStop,
            lineManager,
        )
        var subtitle = TextUtil.replacePlaceholders(
            subtitleTemplate,
            line,
            currentStop,
            null,
            nextStop,
            terminusStop,
            lineManager,
        )

        title = ChatColor.translateAlternateColorCodes('&', title)
        subtitle = ChatColor.translateAlternateColorCodes('&', subtitle)
        passenger.sendTitle(title, subtitle, 0, 1000000, 0)

        var actionbarTemplate = plugin.configFacade.getWaitingActionbar()
        if (customTitle != null && customTitle.containsKey("actionbar")) {
            actionbarTemplate = customTitle.getValue("actionbar")
        }

        if (actionbarTemplate.contains("{countdown}")) {
            startCountdownActionbar(passenger, minecart, line, "waiting", currentStop, null, nextStop, terminusStop)
        } else {
            var actionbarText = TextUtil.replacePlaceholders(
                actionbarTemplate,
                line,
                currentStop,
                null,
                nextStop,
                terminusStop,
                lineManager,
            )
            actionbarText = ChatColor.translateAlternateColorCodes('&', actionbarText)
            passenger.spigot().sendMessage(ChatMessageType.ACTION_BAR, *TextComponent.fromLegacyText(actionbarText))
        }
    }

    private fun startCountdownActionbar(
        passenger: Player?,
        minecart: Minecart?,
        line: Line,
        infoType: String,
        mainStop: Stop,
        prevStop: Stop?,
        nextStop: Stop?,
        terminusStop: Stop?,
    ) {
        if (passenger == null || !passenger.isOnline) {
            return
        }

        var actionbarTemplate = ""
        when (infoType) {
            "waiting" -> actionbarTemplate = plugin.configFacade.getWaitingActionbar()
            "departure" -> actionbarTemplate = plugin.configFacade.getDepartureActionbar()
            "arrive_stop", "terminal_stop" -> {
                // No default actionbar for these yet, wait for implementation if needed
            }
        }

        val customTitle = mainStop.getCustomTitle(infoType)
        if (customTitle != null && customTitle.containsKey("actionbar")) {
            actionbarTemplate = customTitle.getValue("actionbar")
        }

        val template = actionbarTemplate
        val lineManager = plugin.lineManager
        val durationTicks = if (infoType == "waiting") {
            getWaitingDisplayDurationTicks(minecart)
        } else {
            plugin.configFacade.getCartDepartureDelay()
        }
        val totalSeconds = ceil(durationTicks / 20.0).toInt()

        for (secondsLeft in totalSeconds downTo 0) {
            val delayTicks = (totalSeconds - secondsLeft) * 20L
            scheduleTrainTask(
                minecart,
                Runnable {
                    if (!isPassengerStillRidingTrain(minecart, passenger)) {
                        return@Runnable
                    }

                    var text = template.replace("{countdown}", secondsLeft.toString())
                    text = TextUtil.replacePlaceholders(
                        text,
                        line,
                        mainStop,
                        prevStop,
                        nextStop,
                        terminusStop,
                        lineManager,
                    )
                    text = ChatColor.translateAlternateColorCodes('&', text)
                    passenger.spigot().sendMessage(ChatMessageType.ACTION_BAR, *TextComponent.fromLegacyText(text))
                },
                delayTicks,
                -1,
            )
        }
    }

    private fun startWaitingSound(minecart: Minecart?, passenger: Player?) {
        if (
            !plugin.configFacade.isWaitingSoundEnabled() ||
            plugin.configFacade.getWaitingNotes().isEmpty() ||
            passenger == null
        ) {
            return
        }

        scheduleTrainTask(
            minecart,
            Runnable {
                if (isPassengerStillRidingTrain(minecart, passenger)) {
                    playWaitingSoundOnce(passenger)
                }
            },
            plugin.configFacade.getWaitingInitialDelay().toLong(),
            -1,
        )

        val interval = plugin.configFacade.getWaitingSoundInterval()
        if (interval <= 0) {
            return
        }

        val durationTicks = getWaitingDisplayDurationTicks(minecart)
        val repeatTimes = (durationTicks + interval - 1L) / interval
        for (index in 1..repeatTimes) {
            scheduleTrainTask(
                minecart,
                Runnable {
                    if (isPassengerStillRidingTrain(minecart, passenger)) {
                        playWaitingSoundOnce(passenger)
                    }
                },
                plugin.configFacade.getWaitingInitialDelay().toLong() + interval * index,
                -1,
            )
        }
    }

    private fun getWaitingDisplayDurationTicks(minecart: Minecart?): Long {
        val serviceTrain = findServiceTrain(minecart)
        if (serviceTrain != null && serviceTrain.isWaiting) {
            return serviceTrain.getRemainingDwellTicks(SchedulerUtil.getCurrentTick()).toLong()
        }
        return plugin.configFacade.getCartDepartureDelay()
    }

    private fun isPassengerStillRidingTrain(minecart: Minecart?, passenger: Player?): Boolean {
        if (passenger == null || !passenger.isOnline) {
            return false
        }
        if (passenger.vehicle == minecart) {
            return true
        }
        val serviceTrain = findServiceTrain(minecart)
        return serviceTrain != null && serviceTrain.isPassenger(passenger)
    }

    private fun findServiceTrain(minecart: Minecart?): TrainInstance? {
        val lineServiceManager = plugin.lineServiceManager
        if (minecart == null || lineServiceManager == null) {
            return null
        }
        return lineServiceManager.getTrainByMinecart(minecart.uniqueId)
    }

    private fun scheduleTrainTask(minecart: Minecart?, task: Runnable, delay: Long, period: Long): Any? {
        if (minecart == null) {
            return null
        }
        val movementTask = TrainMovementTask.getTaskFor(minecart)
        if (movementTask != null) {
            return movementTask.scheduleSessionTask(task, delay, period)
        }
        return SchedulerUtil.entityRun(plugin, minecart, task, delay, period)
    }

    private fun playWaitingSoundOnce(passenger: Player?) {
        if (passenger != null && passenger.isOnline) {
            SoundUtil.playNoteSequence(plugin, passenger, plugin.configFacade.getWaitingNotes(), 0)
        }
    }

    private fun showStopInfo(
        passenger: Player?,
        line: Line,
        infoType: String,
        mainStop: Stop?,
        prevStop: Stop?,
        nextStop: Stop?,
        terminusStop: Stop?,
        fadeIn: Int,
        stay: Int,
        fadeOut: Int,
    ) {
        if (passenger == null || !passenger.isOnline || mainStop == null) {
            return
        }

        val lineManager = plugin.lineManager
        var titleTemplate = ""
        var subtitleTemplate = ""
        var actionbarTemplate = ""

        when (infoType) {
            "arrive_stop" -> {
                titleTemplate = plugin.configFacade.getArriveStopTitle()
                subtitleTemplate = plugin.configFacade.getArriveStopSubtitle()
            }

            "terminal_stop" -> {
                titleTemplate = plugin.configFacade.getTerminalStopTitle()
                subtitleTemplate = plugin.configFacade.getTerminalStopSubtitle()
            }

            "departure" -> {
                titleTemplate = plugin.configFacade.getDepartureTitle()
                subtitleTemplate = plugin.configFacade.getDepartureSubtitle()
                actionbarTemplate = plugin.configFacade.getDepartureActionbar()
            }

            "waiting" -> {
                titleTemplate = plugin.configFacade.getWaitingTitle()
                subtitleTemplate = plugin.configFacade.getWaitingSubtitle()
                actionbarTemplate = plugin.configFacade.getWaitingActionbar()
            }
        }

        val customTitle = mainStop.getCustomTitle(infoType)
        if (customTitle != null) {
            if (customTitle.containsKey("title")) {
                titleTemplate = customTitle.getValue("title")
            }
            if (customTitle.containsKey("subtitle")) {
                subtitleTemplate = customTitle.getValue("subtitle")
            }
            if (customTitle.containsKey("actionbar")) {
                actionbarTemplate = customTitle.getValue("actionbar")
            }
        }

        var title = TextUtil.replacePlaceholders(
            titleTemplate,
            line,
            mainStop,
            prevStop,
            nextStop,
            terminusStop,
            lineManager,
        )
        var subtitle = TextUtil.replacePlaceholders(
            subtitleTemplate,
            line,
            mainStop,
            prevStop,
            nextStop,
            terminusStop,
            lineManager,
        )
        var actionbar = TextUtil.replacePlaceholders(
            actionbarTemplate,
            line,
            mainStop,
            prevStop,
            nextStop,
            terminusStop,
            lineManager,
        )

        title = ChatColor.translateAlternateColorCodes('&', title)
        subtitle = ChatColor.translateAlternateColorCodes('&', subtitle)
        actionbar = ChatColor.translateAlternateColorCodes('&', actionbar)

        passenger.sendTitle(title, subtitle, fadeIn, stay, fadeOut)
        passenger.spigot().sendMessage(ChatMessageType.ACTION_BAR, *TextComponent.fromLegacyText(actionbar))
    }
}
